velox/docs/develop/hash-table.rst

```diff
@@ -0,0 +1,117 @@
+pointers. Each tag occupies 1 byte (only 7 bits are used). 16 tags occupy 16 bytes. Each pointer
+occupies 6 bytes. 16 pointers occupy 48 bytes. There are 16 bytes left unused at the end of the bucket.
+These are referred to as padding.
+
+.. image:: images/hash-table-layout.png
+    :align: center
+
+A hash table is never full. There are always some empty slots. Velox allows HashTable to fill up to ⅞ of
+capacity before resizing.
+On resize HashTable’s capacity doubles.
+
+Individual buckets may be completely empty, partially filled or full. Buckets are filled left to right.
+If a bucket is partially full, then first N tags and N pointers are filled and the rest are free (N < 16).
+
+
+Inserting an entry
+------------------
+
+To insert a new entry we need to figure out which slot to put it in.
+A slot is identified by bucket and offset within the bucket. First, we compute a hash of the entry.
+Then, we compute a tag and a bucket number from the hash.
+
+We use 7 bits of the hash for the tag: bits 32-38 inclusive. We use N bits of the hash for the bucket
+starting from bit 8.
+
+Given hash number
+
+01011011 11001010 01111100 0\ **1101001** 10110111 1\ **0010100 11111000 1**\ 1001110
+
+We compute the tag as **11101001** and bucket offset as 1,374,336 (00000000 00000000 00000000 00000000
+00000000 00010100 11111000 10000000).
+Bucket offset is the number of bytes from the start of the hash table.
+
+.. code-block:: c++
+
+    bucket offset = bucket number * 128
+    bucket number = bucket offset / 128
+
+We use a bucket offset to get to the bucket. Here we have a few possibilities. The bucket might be empty.
+
+.. image:: images/hash-table-insert1.png
+    :align: center
+
+In this case, we simply insert our entry into the first slot of the bucket.
+We store only the tag (7 bits of the hash value) in the hash table. The hash number itself is not stored.
+The pointer refers to memory outside of the HashTable where the complete value is stored.
+This is usually a row in a RowContainer. The HashTable can be viewed as an index on top of a RowContainer
+that helps locate records with matching keys faster. HashTable itself doesn’t store data or keys.
+
+.. image:: images/hash-table-insert2.png
+    :align: center
+
+Another possibility is a bucket partially filled (for example, has one slot occupied like above).
+In this case there is a possibility that the new entry is a duplicate of the one already stored.
+Hence, we compare the tag of the new entry with tags stored in the bucket.
+If none match, this entry is not a duplicate, so we store it in the next available slot in the bucket.
+
+.. image:: images/hash-table-insert3.png
+    :align: center
+
+However, if one or more existing tags match the tag of the new entry, we follow the pointers to compare
+the keys to determine whether there is a match or not. If there is no match we insert a new entry.
+Otherwise, we drop it.
```

Comment:
Wouldn't this depend on whether the HashTable allows rows with duplicate keys or not ? We do allow duplicates in HashJoin right ? Or is there something else in play here ?

Reply:
This is a bit tricky. Hash table by definition doesn't allow duplicates. Hence all entries in a hash table are unique. However, you are correct that hash table is used for cases where build side of the join has duplicates. One way to think about this is that hash table stores unique keys. For each set of unique keys it has a pointer to a list of rows with matching keys. The "pointer to a list of rows" is implemented as a pointer to the first row in the list. That row may have a link to next row, which may  itself have a link to the next row and so on. There is a special field in RowContainer that allows storing a pointer to the next row (nextOffset_). 

```
template <bool ignoreNullKeys>
void HashTable<ignoreNullKeys>::pushNext(char* row, char* next) {
  if (nextOffset_ > 0) {
    hasDuplicates_ = true;
    auto previousNext = nextRow(row);
    nextRow(row) = next;
    nextRow(next) = previousNext;
  }
}
```

Reply:
@mbasmanova : Yes, this is fair. Thanks for the explanation.